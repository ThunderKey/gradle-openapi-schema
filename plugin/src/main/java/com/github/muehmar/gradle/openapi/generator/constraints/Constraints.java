package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class Constraints {
  private final Min min;
  private final Max max;
  private final DecimalMin decimalMin;
  private final DecimalMax decimalMax;
  private final Size size;
  private final Pattern pattern;
  private final Email email;

  private Constraints(
      Min min,
      Max max,
      DecimalMin decimalMin,
      DecimalMax decimalMax,
      Size size,
      Pattern pattern,
      Email email) {
    this.min = min;
    this.max = max;
    this.decimalMin = decimalMin;
    this.decimalMax = decimalMax;
    this.size = size;
    this.pattern = pattern;
    this.email = email;
  }

  public static Constraints empty() {
    return new Constraints(null, null, null, null, null, null, null);
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public Constraints withMin(Min min) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withMax(Max max) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withSize(Size size) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public void onMin(Consumer<Min> onMin) {
    Optional.ofNullable(min).ifPresent(onMin);
  }

  public void onMax(Consumer<Max> onMax) {
    Optional.ofNullable(max).ifPresent(onMax);
  }

  public void onDecimalMin(Consumer<DecimalMin> onDecimalMin) {
    Optional.ofNullable(decimalMin).ifPresent(onDecimalMin);
  }

  public void onDecimalMax(Consumer<DecimalMax> onDecimalMax) {
    Optional.ofNullable(decimalMax).ifPresent(onDecimalMax);
  }

  public void onSize(Consumer<Size> onSize) {
    Optional.ofNullable(size).ifPresent(onSize);
  }

  public void onPattern(Consumer<Pattern> onPattern) {
    Optional.ofNullable(pattern).ifPresent(onPattern);
  }

  public void onEmail(Consumer<Email> onEmail) {
    Optional.ofNullable(email).ifPresent(onEmail);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Constraints that = (Constraints) o;
    return Objects.equals(min, that.min)
        && Objects.equals(max, that.max)
        && Objects.equals(size, that.size)
        && Objects.equals(pattern, that.pattern)
        && Objects.equals(email, that.email);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max, size, pattern, email);
  }

  @Override
  public String toString() {
    return "Constraints{"
        + "min="
        + min
        + ", max="
        + max
        + ", size="
        + size
        + ", pattern="
        + pattern
        + ", email="
        + email
        + '}';
  }
}
