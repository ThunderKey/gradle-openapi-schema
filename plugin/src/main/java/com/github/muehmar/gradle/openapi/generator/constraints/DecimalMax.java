package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

public class DecimalMax {
  private final String value;
  private final boolean inclusiveMax;

  public DecimalMax(String value, boolean inclusiveMax) {
    this.value = value;
    this.inclusiveMax = inclusiveMax;
  }

  public static DecimalMax inclusive(String value) {
    return new DecimalMax(value, true);
  }

  public DecimalMax exclusiveMax() {
    return new DecimalMax(value, false);
  }

  public String getValue() {
    return value;
  }

  public boolean isInclusiveMax() {
    return inclusiveMax;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    DecimalMax that = (DecimalMax) o;
    return inclusiveMax == that.inclusiveMax && Objects.equals(value, that.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value, inclusiveMax);
  }

  @Override
  public String toString() {
    return "DecimalMax{" + "value='" + value + '\'' + ", inclusiveMax=" + inclusiveMax + '}';
  }
}
