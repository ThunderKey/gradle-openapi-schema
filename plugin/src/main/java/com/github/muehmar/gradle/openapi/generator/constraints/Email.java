package com.github.muehmar.gradle.openapi.generator.constraints;

/** Email constraint for a property */
@SuppressWarnings("java:S2094")
public class Email {}
