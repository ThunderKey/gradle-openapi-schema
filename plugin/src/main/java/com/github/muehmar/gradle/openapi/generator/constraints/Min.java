package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

/** Minimum constraint of a property */
public class Min {
  private final long value;

  public Min(long value) {
    this.value = value;
  }

  public long getValue() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Min min1 = (Min) o;
    return value == min1.value;
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return "Min{" + "value=" + value + '}';
  }
}
