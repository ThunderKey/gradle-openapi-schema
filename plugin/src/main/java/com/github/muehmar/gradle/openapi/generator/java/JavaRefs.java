package com.github.muehmar.gradle.openapi.generator.java;

public class JavaRefs {
  private JavaRefs() {}

  public static final String LOCAL_DATE = "java.time.LocalDate";
  public static final String OPTIONAL = "java.util.Optional";
}
