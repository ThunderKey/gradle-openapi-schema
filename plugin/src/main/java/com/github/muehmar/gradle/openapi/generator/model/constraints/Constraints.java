package com.github.muehmar.gradle.openapi.generator.model.constraints;

import com.github.muehmar.gradle.openapi.util.Optionals;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
@PojoBuilder
public class Constraints {
  Optional<Min> min;
  Optional<Max> max;
  Optional<DecimalMin> decimalMin;
  Optional<DecimalMax> decimalMax;
  Optional<Size> size;
  Optional<Pattern> pattern;
  Optional<Email> email;
  Optional<PropertyCount> propertyCount;
  Optional<MultipleOf> multipleOf;

  public static Constraints empty() {
    return new Constraints(
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty());
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public static Constraints ofDecimalMinAndMax(DecimalMin decimalMin, DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMin(decimalMin).withDecimalMax(decimalMax);
  }

  public static Constraints ofPropertiesCount(PropertyCount propertyCount) {
    return ConstraintsBuilder.create().andOptionals().propertyCount(propertyCount).build();
  }

  public static Constraints ofMultipleOf(MultipleOf multipleOf) {
    return ConstraintsBuilder.create().andOptionals().multipleOf(multipleOf).build();
  }

  public Constraints withMin(Min min) {
    return new Constraints(
        Optional.ofNullable(min),
        max,
        decimalMin,
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withMax(Max max) {
    return new Constraints(
        min,
        Optional.ofNullable(max),
        decimalMin,
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withSize(Size size) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        Optional.ofNullable(size),
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        size,
        Optional.ofNullable(pattern),
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        size,
        pattern,
        Optional.ofNullable(email),
        propertyCount,
        multipleOf);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(
        min,
        max,
        Optional.ofNullable(decimalMin),
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(
        min,
        max,
        decimalMin,
        Optional.ofNullable(decimalMax),
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints and(Constraints other) {
    return new Constraints(
        Optionals.or(min, min),
        Optionals.or(max, other.max),
        Optionals.or(decimalMin, other.decimalMin),
        Optionals.or(decimalMax, other.decimalMax),
        Optionals.or(size, other.size),
        Optionals.or(pattern, other.pattern),
        Optionals.or(email, other.email),
        Optionals.or(propertyCount, other.propertyCount),
        Optionals.or(multipleOf, other.multipleOf));
  }

  public Optional<Min> getMin() {
    return min;
  }

  public Optional<Max> getMax() {
    return max;
  }

  public Optional<DecimalMin> getDecimalMin() {
    return decimalMin;
  }

  public Optional<DecimalMax> getDecimalMax() {
    return decimalMax;
  }

  public Optional<Size> getSize() {
    return size;
  }

  public Optional<Pattern> getPattern() {
    return pattern;
  }

  public Optional<Email> getEmail() {
    return email;
  }

  public <R> Optional<R> onMinFn(Function<Min, R> onMin) {
    return min.map(onMin);
  }

  public <R> Optional<R> onMaxFn(Function<Max, R> onMax) {
    return max.map(onMax);
  }

  public <R> Optional<R> onDecimalMinFn(Function<DecimalMin, R> onDecimalMin) {
    return decimalMin.map(onDecimalMin);
  }

  public <R> Optional<R> onDecimalMaxFn(Function<DecimalMax, R> onDecimalMax) {
    return decimalMax.map(onDecimalMax);
  }

  public <R> Optional<R> onSizeFn(Function<Size, R> onSize) {
    return size.map(onSize);
  }

  public <R> Optional<R> onPatternFn(Function<Pattern, R> onPattern) {
    return pattern.map(onPattern);
  }
}
