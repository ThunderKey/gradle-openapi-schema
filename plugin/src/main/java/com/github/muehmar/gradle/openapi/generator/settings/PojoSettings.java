package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
