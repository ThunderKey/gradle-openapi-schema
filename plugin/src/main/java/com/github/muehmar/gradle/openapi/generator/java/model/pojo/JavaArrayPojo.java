package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaArrayType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ArrayPojo;
import com.github.muehmar.gradle.openapi.generator.model.type.ArrayType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaArrayPojo implements JavaPojo {
  private final PojoName name;
  private final String description;
  private final JavaType itemType;
  private final Constraints constraints;
  private final JavaPojoMember arrayPojoMember;

  private JavaArrayPojo(
      PojoName name,
      String description,
      JavaType itemType,
      Constraints constraints,
      JavaPojoMember arrayPojoMember) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.itemType = itemType;
    this.constraints = constraints;
    this.arrayPojoMember = arrayPojoMember;
  }

  public static JavaArrayPojo wrap(ArrayPojo arrayPojo, TypeMappings typeMappings) {
    final JavaType itemType = JavaType.wrap(arrayPojo.getItemType(), typeMappings);
    final JavaPojoMember arrayPojoMember = createItemTypeMember(arrayPojo, typeMappings);
    return new JavaArrayPojo(
        arrayPojo.getName(),
        arrayPojo.getDescription(),
        itemType,
        arrayPojo.getConstraints(),
        arrayPojoMember);
  }

  private static JavaPojoMember createItemTypeMember(
      ArrayPojo arrayPojo, TypeMappings typeMappings) {
    final ArrayType arrayType =
        ArrayType.ofItemType(arrayPojo.getItemType()).withConstraints(arrayPojo.getConstraints());
    final JavaArrayType javaArrayType = JavaArrayType.wrap(arrayType, typeMappings);
    final Name name = Name.ofString("value");
    return JavaPojoMember.of(
        name,
        arrayPojo.getDescription(),
        javaArrayType,
        Necessity.REQUIRED,
        Nullability.NOT_NULLABLE);
  }

  @Override
  public PojoName getName() {
    return name;
  }

  @Override
  public String getDescription() {
    return description;
  }

  public JavaPojoMember getArrayPojoMember() {
    return arrayPojoMember;
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo,
      Function<JavaComposedPojo, T> onComposedPojo,
      Function<JavaFreeFormPojo, T> onFreeFormPojo) {
    return onArrayPojo.apply(this);
  }
}
