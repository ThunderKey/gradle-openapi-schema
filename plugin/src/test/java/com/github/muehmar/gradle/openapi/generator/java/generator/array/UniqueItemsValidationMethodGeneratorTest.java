package com.github.muehmar.gradle.openapi.generator.java.generator.array;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaArrayPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class UniqueItemsValidationMethodGeneratorTest {
  private Expect expect;

  @Test
  void generate_when_noUniqueItemsConstraint_then_noOutput() {
    final Generator<JavaArrayPojo, PojoSettings> generator =
        UniqueItemsValidationMethodGenerator.generator();

    final Writer writer =
        generator.generate(
            JavaPojos.arrayPojo(),
            TestPojoSettings.defaultSettings().withEnableValidation(true),
            Writer.createDefault());

    assertEquals("", writer.asString());
  }

  @Test
  void generate_when_validationDisabled_then_noOutput() {
    final Generator<JavaArrayPojo, PojoSettings> generator =
        UniqueItemsValidationMethodGenerator.generator();
    final Writer writer =
        generator.generate(
            JavaPojos.arrayPojo(Constraints.ofUniqueItems(true)),
            TestPojoSettings.defaultSettings().withEnableValidation(false),
            Writer.createDefault());

    assertEquals("", writer.asString());
  }

  @Test
  @SnapshotName("uniqueItemsConstraint")
  void generate_when_uniqueItemsConstraint_then_correctOutput() {
    final Generator<JavaArrayPojo, PojoSettings> generator =
        UniqueItemsValidationMethodGenerator.generator();
    final Writer writer =
        generator.generate(
            JavaPojos.arrayPojo(Constraints.ofUniqueItems(true)),
            TestPojoSettings.defaultSettings().withEnableValidation(true),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_HASH_SET::equals));
    expect.toMatchSnapshot(writer.asString());
  }
}
