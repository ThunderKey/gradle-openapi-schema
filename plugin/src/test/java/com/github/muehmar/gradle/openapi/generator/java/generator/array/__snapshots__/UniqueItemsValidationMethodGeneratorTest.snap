uniqueItemsConstraint=[
@AssertTrue(message = "PosologyDto does not contain unique items")
private boolean isAllUniqueItems() {
  return new HashSet<>(value).size() == value.size();
}
]